### ISSUE TITLE
将segnext的动态图导出为静态图模型，静态图模型出现了严重的推理错误

### ISSUE THREAD

**Issue by ghost** [PaddlePaddle/PaddleSeg#3289]
### 问题确认 Search before asking

- [X] 我已经查询[历史issue](https://github.com/PaddlePaddle/PaddleSeg/issues)(包括open与closed)，没有发现相似的bug。I have searched the [open and closed issues](https://github.com/PaddlePaddle/PaddleSeg/issues) and found no similar bug report.

### Bug描述 Describe the Bug

将segnext的动态图导出为静态图模型，静态图模型出现了严重的推理错误，静态图推理出来的图片全是背景，而动态图推理的图片是正常的

### 复现环境 Environment

AI Stduio
python3.7
paddleseg2.8
paddle2.4.0

### Bug描述确认 Bug description confirmation

- [X] 我确认已经提供了Bug复现步骤、代码改动说明、以及环境信息，确认问题是可以复现的。I confirm that the bug replication steps, code change instructions, and environment information have been provided, and the problem can be reproduced.

### 是否愿意提交PR？ Are you willing to submit a PR?

- [ ] 我愿意提交PR！I'd like to help by submitting a PR!

**Comment by ghost** [comment]
%cd ~
%cd PaddleSeg
!python tools/predict.py \
       --config configs/segnext_s_1024x1024.yml \
       --model_path output/segnext_s_1024x1024/best_model/model.pdparams \
       --image_path /home/aistudio/work/car_data_2022/JPEGImages \
       --save_dir /home/aistudio/work/result


%cd ~
%cd PaddleSeg
!python tools/export.py --config configs/segnext_s_1024x1024.yml \
                  --model_path output/segnext_s_1024x1024/best_model/model.pdparams \
                  --save_dir /home/aistudio/test_fps/model/segnext_s_1024x1024 \
                  --input_shape 1 3 870 490

**Comment by Bobholamovic** [comment]
你好，经过测试我无法复现这个问题（使用的环境也是Python3.7+PaddleSeg 2.8+PaddlePaddle2.4.0）。我建议着重检查一下推理时模型是否正确载入，以及导出时的`--input_shape`是否正确（顺序为`B C H W`）~

**Comment by ghost** [comment]
> 静态图推理时模型正确载入了，无任何报错和警告信息。导出时的input_shape 1 3 870 490也无任何问题。我换成其他静态图模型，比如ocrnet、deeplabv3p和rtformer进行推理出的图，预测结果跟标签差不多，但是换成这个segnext静态图推理跟标签差异非常大，可视化后几乎只有1个类别（标签有4个类别），但segnext的动态图推理却很正常，跟标签差不多。所以我怀疑segnext导出后的静态图模型有bug

**Comment by ghost** [comment]
> 你好，经过测试我无法复现这个问题（使用的环境也是Python3.7+PaddleSeg 2.8+PaddlePaddle2.4.0）。我建议着重检查一下推理时模型是否正确载入，以及导出时的`--input_shape`是否正确（顺序为`B C H W`）~

最近在ai stduio打一个图像分割的比赛，基于训练时都有0.75左右的miou，唯独这个segnext模型，导出后的静态图模型分数非常低,只有0.005miou左右，而其他的模型导出后至少有0.6以上的miou。经过我动态图执行predict.py和静态图推理相同的图片，灰度图经可视化后验证了我的猜想。

静态图推理时模型正确载入了，无任何报错和警告信息。导出时的input_shape 1 3 870 490也无任何问题。我换成其他静态图模型，比如ocrnet、deeplabv3p和rtformer进行推理出的图，预测结果跟标签差不多，但是换成这个segnext静态图推理跟标签差异非常大，可视化后几乎只有1个类别（标签有4个类别），但segnext的动态图推理却很正常，跟标签差不多。所以我怀疑segnext导出后的静态图模型有bug

**Comment by Bobholamovic** [comment]
> 静态图推理时模型正确载入了，无任何报错和警告信息。导出时的input_shape 1 3 870 490也无任何问题。我换成其他静态图模型，比如ocrnet、deeplabv3p和rtformer进行推理出的图，预测结果跟标签差不多，但是换成这个segnext静态图推理跟标签差异非常大，可视化后几乎只有1个类别（标签有4个类别），但segnext的动态图推理却很正常，跟标签差不多。所以我怀疑segnext导出后的静态图模型有bug

请问是否测试过不同的数据呀？测试多张图像，这个问题是否能够稳定复现哦？

**Comment by ghost** [comment]
> > 静态图推理时模型正确载入了，无任何报错和警告信息。导出时的input_shape 1 3 870 490也无任何问题。我换成其他静态图模型，比如ocrnet、deeplabv3p和rtformer进行推理出的图，预测结果跟标签差不多，但是换成这个segnext静态图推理跟标签差异非常大，可视化后几乎只有1个类别（标签有4个类别），但segnext的动态图推理却很正常，跟标签差不多。所以我怀疑segnext导出后的静态图模型有bug
> 
> 请问是否测试过不同的数据呀？测试多张图像，这个问题是否能够稳定复现哦？

测试过很多不同的数据，都是segnext静态图模型推理后的图只有1类，而动态图没问题。静态图推理代码没问题。我认为能稳定复现

**Comment by Bobholamovic** [comment]
考虑到之前确实没有做过segnext模型在Cityscapes以外数据集上的验证，我补充了一些实验。

1. 在Cityscapes数据集（19类，导出尺寸为[1, 3, 1024, 2048]，PaddleSeg提供的segnext模型的默认配置）上训练，segnext模型的动态图推理（左）和静态图推理（右）结果分别如下（给出2组结果）：
![image](https://github.com/PaddlePaddle/PaddleSeg/assets/21275753/6ce43a23-b767-426f-ab04-2c91cec03f22)
![image](https://github.com/PaddlePaddle/PaddleSeg/assets/21275753/114e31ec-a644-47c7-9b90-1b5cc227efe3)

2. 在optic_disc_seg数据集上（2类，导出尺寸为[1, 3, 512, 512]）训练，segnext模型的动态图推理（左）和静态图推理（右）结果分别如下：
![image](https://github.com/PaddlePaddle/PaddleSeg/assets/21275753/c20d96ff-891a-4065-bc01-a62bb50bf1b2)

*注：动静态图推理均使用GPU，没有开启TensorRT。测试图像均为随机选取。*

可以看出，动静态图推理结果基本一致，我还是没能复现你提到的问题……

我建议首先检查一下使用的配置文件和脚本是否无误（尤其是导出时使用的配置文件需要和动态图推理使用的吻合）。如果确认无误，可以尝试对PaddleSeg提供的Cityscapes数据集上预训练的模型进行导出和推理，排查原因是否和本地环境相关。

**Comment by ghost** [comment]
> 考虑到之前确实没有做过segnext模型在Cityscapes以外数据集上的验证，我补充了一些实验。
> 
> 1. 在Cityscapes数据集（19类，导出尺寸为[1, 3, 1024, 2048]，PaddleSeg提供的segnext模型的默认配置）上训练，segnext模型的动态图推理（左）和静态图推理（右）结果分别如下（给出2组结果）：
>    ![image](https://user-images.githubusercontent.com/21275753/240895802-6ce43a23-b767-426f-ab04-2c91cec03f22.png)
>    ![image](https://user-images.githubusercontent.com/21275753/240896737-114e31ec-a644-47c7-9b90-1b5cc227efe3.png)
> 2. 在optic_disc_seg数据集上（2类，导出尺寸为[1, 3, 512, 512]）训练，segnext模型的动态图推理（左）和静态图推理（右）结果分别如下：
>    ![image](https://user-images.githubusercontent.com/21275753/240894960-c20d96ff-891a-4065-bc01-a62bb50bf1b2.png)
> 
> _注：动静态图推理均使用GPU，没有开启TensorRT。测试图像均为随机选取。_
> 
> 可以看出，动静态图推理结果基本一致，我还是没能复现你提到的问题……
> 
> 我建议首先检查一下使用的配置文件和脚本是否无误（尤其是导出时使用的配置文件需要和动态图推理使用的吻合）。如果确认无误，可以尝试对PaddleSeg提供的Cityscapes数据集上预训练的模型进行导出和推理，排查原因是否和本地环境相关。

非常感谢您的回答，以下是我使用paddleseg的过程：
#开始训练
%cd ~
%cd PaddleSeg
!python tools/train.py \
       --config configs/segnext_s_1024x1024.yml \
       --do_eval \
       --use_vdl \
       --num_workers 4 \
       --save_interval 500 \
       --save_dir output/segnext_s_1024x1024 \
       --resume_model output/segnext_s_1024x1024/iter_77000 \ 

%cd ~
%cd PaddleSeg
!python tools/export.py --config configs/segnext_s_1024x1024.yml \
                  --model_path output/segnext_s_1024x1024/best_model/model.pdparams \
                  --save_dir /home/aistudio/test_fps/model/segnext_s_1024x1024 \
                  --input_shape 1 3 870 490

%cd ~
%cd PaddleSeg
!python tools/predict.py \
       --config configs/segnext_s_1024x1024.yml \
       --model_path output/segnext_s_1024x1024/best_model/model.pdparams \
       --image_path /home/aistudio/work/car_data_2022/JPEGImages \
       --save_dir /home/aistudio/work/result

%cd ~
%cd test_fps
!python predict.py test_model.txt result.json

我认为没什么问题，数据集是全国第17届智能汽车竞赛完全模型组线上赛的。训练确实能达到0.77miou左右。

可否透入一下您的静态图推理代码。我的成绩提交代码交其他模型确实没问题，但交这个segnext却有问题

**Comment by ghost** [comment]
> 
训练用的配置文件：segnext_s_1024x1024.yml


batch_size: 14
iters: 80000

train_dataset:
  type: Dataset
  dataset_root: /home/aistudio/work/car_data_2022
  train_path: /home/aistudio/work/car_data_2022/train.txt
  num_classes: 4
  transforms:
    - type: RandomHorizontalFlip
    - type: RandomVerticalFlip
    - type: RandomNoise
    - type: RandomBlur
    - type: RandomRotation
    - type: Resize
      target_size: [870,490]
      interp: 'NEAREST'
    - type: Normalize
      mean: [0.485, 0.456, 0.406]
      std: [0.229, 0.224, 0.225]
    - type: RandomErasings
      prob: 0.5
      scale: [0.02, 0.4]
      ratio: [0.3, 3.3]
      value: 'random'
      inplace: False
  mode: train

val_dataset:
  type: Dataset
  dataset_root: /home/aistudio/work/car_data_2022
  val_path: /home/aistudio/work/car_data_2022/val.txt
  num_classes: 4
  transforms:
    - type: Resize
      target_size: [870,490]
      interp: 'NEAREST'
    - type: Normalize
      mean: [0.485, 0.456, 0.406]
      std: [0.229, 0.224, 0.225]
  mode: val

optimizer:
  type: AdamW
  weight_decay: 0.01
  custom_cfg:
    - name: head
      lr_mult: 10.0
    - name: norm
      weight_decay_mult: 0.0


loss:
  types:
    - type: MixedLoss
      losses:
        - type: LaneCrossEntropyLoss
          weights: [0.5, 1.0, 1.0, 1.0]    
        - type: LovaszSoftmaxLoss
      coef: [1, 0.2]
  coef: [1]

lr_scheduler:
  type: PolynomialDecay
  learning_rate: 0.00006
  end_lr: 0.0
  power: 1.0

model:
  type: SegNeXt
  backbone:
    type: MSCAN_S
    pretrained: https://paddleseg.bj.bcebos.com/dygraph/backbone/mscan_s_imagenet_pretrained.zip
  num_classes: 4
  decoder_cfg:
    channels: 256
    ham_channels: 256
    ham_kwargs:
      MD_R: 16
    dropout_rate: 0.1
    align_corners: False

**Comment by Bobholamovic** [comment]
从指令和配置文件上看的话，我只能发现一个小问题：PaddleSeg中`Resize`要求以`[W, H]`指定形状（和OpenCV一致），而`--input_shape`指定的是输入tensor的形状所以顺序是`B C H W`。因此，你的配置文件和执行的指令中一定有一个是不正确的——正确的组合要么是`[870, 490]`和`--input_shape 1 3 490 870`，要么是`[490, 870]`和`--input_shape 1 3 870 490`。

不过，考虑到你的其它模型静态图没问题，我们所面临的主要问题也可能不是这个原因导致的。建议在修复这个错误之后可以尝试对PaddleSeg提供的Cityscapes数据集上预训练的模型进行导出和推理，排查原因是否和本地环境相关。

我在实验中使用PaddleSeg建议的标准方式执行静态图推理，例如：
```bash
python deploy/python/infer.py --config sta/deploy.yaml --image_path data/cityscapes/leftImg8bit/train/cologne/cologne_000012_000019_leftImg8bit.png --save_dir sta
```
